LSM for v0.47.x

xlab opened the pull request:
Better view commit by commit, the first one is the actual implementation.

Commit:
feat: LSM implementation on top of cosmos sdk v0.47.x

This implementation includes validator bond and global liquid staking cap.

Commit:
chore: update buf files to point to buf.build/persistence/cosmos-sdk-lsm

Commit:
chore: re-gen types from proto using buf

Commit:
chore: re-gen interface mocks

Commit:
fix: update SDK unit/integration tests to support LSM modifications

xlab commented:
Some additional E2E tests fixes are WIP

Commit:
chore: run make proto-all to reformat some protos.

ajeet97 commented on tests/e2e/staking/suite.go:
```diff
@@ -68,24 +68,14 @@ func (s *E2ETestSuite) SetupSuite() {
+	_, err = s.network.WaitForHeight(1)
 	s.Require().NoError(err)
-	s.Require().NoError(val.ClientCtx.Codec.UnmarshalJSON(out.Bytes(), &txRes))
-	s.Require().Equal(uint32(0), txRes.Code)
```
why is this (and below) check removed?

xlab replied:
This part was borked, I replaced with original test suite init - wasn't supposed to be changed.

ajeet97 commented on tests/e2e/staking/suite.go:
```diff
@@ -68,24 +68,14 @@ func (s *E2ETestSuite) SetupSuite() {
+	_, err = MsgUnbondExec(val.ClientCtx, val.Address, val.ValAddress, unbond)
 	s.Require().NoError(err)
+
+	_, err = s.network.WaitForHeight(1)
```
wait for height 2? (height 1 has already reached)

xlab replied:
This part was borked, I replaced with original test suite init - wasn't supposed to be changed.

ajeet97 commented on x/distribution/keeper/msg_server.go:
```diff
@@ -158,3 +158,73 @@ func (k msgServer) CommunityPoolSpend(goCtx context.Context, req *types.MsgCommu
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
```
`EventTypeMessage` with `AttributeKeyModule` and `AttributeKeySender` can be removed. [ref](https://github.com/cosmos/cosmos-sdk/blob/release/v0.47.x/UPGRADING.md#all)

xlab replied:
Removed 2 occurrences from the new code

ajeet97 commented on x/distribution/types/msg.go:
```diff
@@ -24,6 +27,11 @@ var (
+	_ sdk.Msg = (*MsgWithdrawTokenizeShareRecordReward)(nil)
+	_ sdk.Msg = (*MsgWithdrawAllTokenizeShareRecordReward)(nil)
+
+	_ legacytx.LegacyMsg = (*MsgWithdrawTokenizeShareRecordReward)(nil)
```
why do we need `LegacyMsg` implementation?

xlab replied:
Because simulations in SDK are currently written in a way that expects messages as `legacytx.LegacyMsg` there:
https://github.com/cosmos/cosmos-sdk/blob/10b3db2318f0df1521d00913bfad82c9c0e6d784/types/simulation/types.go#L91-L93

They have already rewritten that part for the next version (whenever it comes).

ajeet97 commented on x/staking/simulation/operations.go:
```diff
@@ -574,3 +617,212 @@ func SimulateMsgBeginRedelegate(ak types.AccountKeeper, bk types.BankKeeper, k *
+				balance := bk.GetBalance(ctx, acc.Address, record.GetShareTokenDenom())
+				if balance.Amount.IsPositive() {
+					simAccount = acc
+					transferRecord = record
```
chosen `transferRecord` may have different owner than `simAccount` so it will fail randomly with err `ErrNotTokenizeShareRecordOwner`

xlab replied:
Added an `if` precheck that validates the owner

puneet2019 commented:
lgtm, just Ajeet's comments

Commit:
chore: pick LSM's upstream changes

PR 86 https://github.com/iqlusioninc/liquidity-staking-module/pull/86
PR 87 https://github.com/iqlusioninc/liquidity-staking-module/pull/87
PR 108 https://github.com/iqlusioninc/liquidity-staking-module/pull/108
PR 110 https://github.com/iqlusioninc/liquidity-staking-module/pull/110

Commit:
chore: re-gen protos

Commit:
chore: get LSM unit/integration tests from upstream and adapt for SDK 47

Commit:
fix: address ENUM_VALUE_PREFIX and ENUM_ZERO_VALUE_SUFFIX proto linting errors

Commit:
chore: address PR comments

xlab commented:
@ajeet97 ok this is final, please check if your comments were addressed, and let's merge.
